### ISSUE TITLE
[@typescript-eslint/no-use-before-define] false positive when using forwardRef(() => Class)

### ISSUE THREAD

**Issue by Res42** [angular-eslint/angular-eslint#504]
**Description and reproduction of the issue**

I don't know if this is the best place to report this issue or in [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint). It is not exactly an Angular specific issue, but it includes an Angular feature so I thought I will start here. If this is out of scope for you I'll take it to them.

Description:

[@typescript-eslint/no-use-before-define](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-use-before-define.md) reports a false positive when using `forwardRef(() => Class)`.

Reproduction:

1. Turn on the rule [@typescript-eslint/no-use-before-define](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-use-before-define.md) rule with (at least) `{ "classes": true }`.
```jsonc
{
    // note you must disable the base rule as it can report incorrect errors
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": ["error", { "classes": true }],
}
```
2. Classes inside `forwardRef`s will trigger the rule, but it shouldn't since this is exactly the use case of `forwardRef`.
```ts
import { Component, forwardRef, Provider } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

export const CUSTOM_VALUE_ACCESSOR: Provider = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => CustomInputComponent),
  //                            ~~~~~~~~~~~~~~~~~~~~
  //                            'CustomInputComponent' was used before it was defined. eslint(@typescript-eslint/no-use-before-define)
  multi: true,
};

@Component({
  selector: 'app-custom-input',
  templateUrl: './custom-input.component.html',
  providers: [CUSTOM_VALUE_ACCESSOR],
})
export class CustomInputComponent {
  // ...
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint`        | `4.24.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.26.0` |
| `node`                             | `14.16.0` |

```sh
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 12.0.1
Node: 14.16.0
Package Manager: npm 6.14.11
OS: win32 x64

Angular: 12.0.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1200.1
@angular-devkit/build-angular   12.0.1
@angular-devkit/core            12.0.1
@angular-devkit/schematics      12.0.1
@schematics/angular             12.0.1
ng-packagr                      12.0.0
rxjs                            6.6.7
typescript                      4.2.4
```

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest supported version of the packages and checked my `ng version` output per the instructions given here.

**Comment by JamesHenry** [comment]
@Res42 Certainly it would be typescript-eslint that would be the appropriate project to report such things, angular-eslint is only concerned with tooling on top of typescript-esint which is specific to the Angular ecosystem - it does not influence the behaviour of typescript-eslint rules.

Nevertheless, as you yourself point out the purpose of the `forwardRef()` utility is to allow you to use things before they are defined. The purpose of `no-use-before-define` is to warn you when you do just that.

Angular's official docs reiterate the behaviour of `forwardRef`:

```ts
class Door {
  lock: Lock;

  // Door attempts to inject Lock, despite it not being defined yet.
  // forwardRef makes this possible.
  constructor(@Inject(forwardRef(() => Lock)) lock: Lock) {
    this.lock = lock;
  }
}

// Only at this point Lock is defined.
class Lock {}

const injector =
    Injector.create({providers: [{provide: Lock, deps: []}, {provide: Door, deps: [Lock]}]});

expect(injector.get(Door) instanceof Door).toBe(true);
expect(injector.get(Door).lock instanceof Lock).toBe(true);
```

The lint rule is therefore behaving as intended, it is telling you you are using something before it is defined, which is its whole purpose.

You, as the author of the code, know that is what you are intending to do in this particular scenario because at a level above the language level (your framework specific stuff) you have a utility.

This IMO is therefore a case where what you want is to use an `eslint-disable` comment for this specific usage. It is rare, specific, and not a language-level concern.

**Comment by Res42** [comment]
I accept your reasoning, but I just want to mention some of my thoughts:

`@typescript-eslint/no-use-before-define` "extends" `no-use-before-define` (and requires `you must disable the base rule as it can report incorrect errors`).
I thought that just like `@typescript-eslint/no-use-before-define` this plugin should "extend" `@typescript-eslint/no-use-before-define` to fix incorrect errors like `forwardRef`. Just as you said:

> angular-eslint is only concerned with tooling on top of typescript-esint which is specific to the Angular ecosystem
